### ISSUE TITLE
Gestion des changements de nomenclature des cases de la déclaration fiscale

### ISSUE THREAD

**Issue by ClaireLeroyIPP** [openfisca/openfisca-france#853]
Après avoir effectuer des tests sur la législation de l'impôt sur le revenu, il y a un problème qui revient souvent sur la définition de variables issues des déclarations d'impôt (2042). **Certaines cases de la déclaration fiscale changent de signification d'une année à l'autre.** 

- Soit la variable initiale que désignait cette case disparait et on reprend le nom de cette case pour désigner une nouvelle variable

Ex: _la case 7UH._
En 2002 : Investissements non professionnel réalisé par une entreprise dans les DOM-TOM
En 2009 : Versement sur un compte épargne codéveloppement
En 2016: Dons et cotisations versés aux partis politiques

- Soit la case englobe désormais plusieurs variables, précédemment définies dans des cases différentes

Ex: _les revenus industriels et commerciaux professionnels imposable (régime du bénéfice réel avec CGA ou viseur)._
Avant 2016 : 
revenus imposables (hors locations meublées) = cases 5KC, 5LC, 5MC => variable _abic_impn_
revenus imposables (locations meublées) =  cases 5HA, 5IA, 5JA => variable _abic_impm_
A partir de 2016 : 
revenus imposables (y compris locations meublées) = cases 5KC, 5LC, 5MC => quelle variable définir dans Openfisca-france ?

----------------

Etant donné qu'on ne peut plus définir de start_date, stop_date pour les variables, il me semble, il faudrait réfléchir à comment gérer ces changements de nomenclature au fil des années. **Plusieurs types de solutions différentes semblent coexister** pour l'instant, ce qui n'est pas forcement très lisible : 

- **Création de différentes variables avec nom de case + années, à partir des cerfa_field**
Ex : il existe une variable _f7uh_ et une variable _f7uh_2007_ (avec chacune un label différent) 
- **Création d'une seule variables avec nom de case à partir des cerfa_field, qui est ensuite appelée par différentes variables intermédiaires dont la formule dépend de l'année.**
Ex: _f7uh_ est utilisée pour calculer _dfppce_ à partir de 2013 (dons de financement de partis) mais aussi pour calculer _ecodev_ en 2009 (compte épargne codéveloppement). Le problème étant que _f7uh_ est labellisé "Dons et cotisations versés aux partis politiques" ...

**Comment by benjello** [comment]
@Anna-Livia @sandcha @fpagnoux : comme le souligne Claire, l'attribut `start_date` est assez important pour les cases de la feuille d'impôt.

Pourrait-on envisager de la réintégrer ? (et pourquoi l'a-t-on virer au fait ?)

Merci !

**Comment by fpagnoux** [comment]
Même avec l'attribut `start-date` et des périodes d'existences différentes, il n'a jamais été possible de créer deux variables avec le même nom. 

Je ne suis pas sûr que ce soit même une bonne idée. Une variable nommée `f7UH` ce n'est déjà pas forcément très clair pour quelqu'un qui n'est pas immergé dans le CERFA de l'IRPP, alors des variables `f7UH` qui changent de signification suivant les années... 

Du coup la piste que je privilégierais c'est de dissocier le nom de la variable du nom de la case CERFA quand le problème se pose. Je crois que l'attribut `cerfa_field` permet déjà d'associer une variable à un champs CERFA sans d'avoir le même nom, non ?

**Comment by fpagnoux** [comment]
Sur la suppression du `start_date` : voir [l'issue](https://github.com/openfisca/openfisca-core/issues/371). De mémoire, c'était redondant avec les start dates des formules.

**Comment by benjello** [comment]
@fpagnoux : je suis d'accord qu'il faut changer le nom de la variable et se rabattre sur `cerfa_field` si ces derniers sont identiques.

Par contre, on aimerait savoir qu'une variable d''input (donc sans formula) n'existait pas avant une certaine date et donc qu'elle renvoit une erreur ou passe encore prenne sa valeur par défaut si elle est appelée avant cette date.

**Comment by benjello** [comment]
@sandcha @fpagnoux : Y a-t-il un moyen de gérer la naissance et la mort de variables pures input ?  

cc @ClaireLeroyIPP @Anna-Livia

**Comment by fpagnoux** [comment]
La mort oui, avec un attribut de variable comme `end = '2012-12-31'`.

> Par contre, on aimerait savoir qu'une variable d''input (donc sans formula) n'existait pas avant une certaine date et donc qu'elle renvoie une erreur ou passe encore prenne sa valeur par défaut si elle est appelée avant cette date.

Dans le comportement actuel, si une variable n'a pas de formule, alors quand on la calcule pour une période:
- Soit un input a été défini, et alors on obtient cet input
- Soit aucun input n'est défini pour la période, et on obtient la valeur par défaut

Donc en soit, il semble que "pas de start date" est équivalent à "start date" + deuxième comportement que tu proposes. À moins que tu ne suggères que la variable retourne la valeur par défaut _même si_ l'utilisateur a défini un input, ce qui me paraît assez contre-intuitif. 

Sur le premier comportement proposé (renvoyer une erreur) pourquoi pas, mais:
- A minima, il faudra aussi renvoyer une erreur si on définit une entrée pour une variable _après_ sa date de fin
- Si on va dans ce sens, est-ce qu'on souhaite aussi que le calcul d'une variable après sa date de fin (et potentiellement avant sa date de début) renvoie une erreur plutôt que la valeur par défaut ? 
    - Si non, on a un comportement inhomogène qui peut être un peu contre-intuitif: j'ai le droit de calculer F74B en 2014, mais pas le droit de la définir en input ? Cette inhomogéniété peut être justifiée, mais il faudrait dans ce cas que la doc soit bien claire sur la question.
    - Si oui, ça veut dire qu'il faut être plus strict dans la rédaction des formules. Quitte à multiplier les dated formulas qui se ressemblent : par exemple si la PAJE disparaît en 2016, alors je dois introduire une nouvelle formule pour _toutes_ les variables qui dépendent de la PAJE, nouvelle formule qui est probablement un copier-coller de l'ancienne à une ligne près, alors qu'avant je pouvais me reposer implicitement sur les valeurs par défauts. Est-ce que c'est dans cette direction qu'on veut aller ? Je n'ai pas forcément d'objection absolue, mais il faut que tout le monde soit bien au clair sur les conséquences.
- Pour les cas d'usages type application web, il faut qu'on soit un minimum flexibles pour éviter les problèmes de productions qui arrivent du jour au lendemain. Donc si les simulations type IPP demandent un mode "strict", il faudra être capable de switcher en mode "flexible" pour l'API Web. C'est possible et justifié si les deux cas d'usages ont des exigences incompatibles, mais ça amène son lot de complexité de potentiels ennuis.

**Comment by Morendil** [comment]
@claireleroy Tu as toutes les fiches de calculs facultatifs en un seul endroit ?

https://www.impots.gouv.fr/portail/files/formulaires/2042/2011/2042_267.pdf
https://www.impots.gouv.fr/portail/files/formulaires/2042/2013/2042_266.pdf
https://www.impots.gouv.fr/portail/files/formulaires/2042/2014/2042_265.pdf
https://www.impots.gouv.fr/portail/files/formulaires/2042/2015/2042_264.pdf
https://www.impots.gouv.fr/portail/files/formulaires/2042/2016/2042_257.pdf
https://www.impots.gouv.fr/portail/files/formulaires/2042/2017/2042_1891.pdf
https://www.impots.gouv.fr/portail/files/formulaires/2042/2018/2042_2324.pdf

**Comment by Morendil** [comment]
Mon avis sur la question: ces inputs ne devraient pas être des variables au sens usuel dans OpenFisca. On ne devrait pas les trouver éparpillés dans le legislation explorer. Idéalement ils seraient indexés, mais renverraient à une page documentaire unique "Déclaration d'impôts de l'année N", qui les listerait avec leurs libellés.

Je verrais bien une évolution de Core pour faciliter ces regroupements. Le plus simple serait sans doute de les lier à des meta-variables déclarées classiquement, en surchargeant set_input pour fournir un comportement spécifique:

```Python
class declaration_2009(Variable):
  set_input = set_input_cerfa
```

L'extension set-input-cerfa nous permettrait de spécifier des inputs comme ceci:

```YAML
input_variables:
  declaration_2009-f7qx: 235.0
```

Resterait à surcharger le __call__ de l'entité FoyerFiscal afin de pouvoir continuer à les appeler sans modifications:
```Python
  f7qx = foyer_fiscal('f7qx',"2009-03")
```
…la période étant suffisante pour cibler la bonne version.

On est pas très loin de quelque chose d'implémentable directement, la seule question il me semble est celle de la représentation interne de ces variables multiples au sein d'une meta-variable; je penche pour un `dict` par simplicité mais je connais encore mal les Holder et Storage et il y a peut-être un loup.

@fpagnoux WDYT ?

**Comment by claireleroy** [comment]
@morendil Pour les fiches de calculs facultatifs de l'impôt, on en a quelques unes pour avant 2006 à l'IPP mais pas tant que ça (2000-2003 essentiellement). Je peux te les envoyer si tu es intéressé. 

Personnellement, j’utilise plutôt les [brochures pratiques de l'IR](https://www.impots.gouv.fr/portail/www2/fichiers/documentation/brochure/ir_2018/files/assets/common/downloads/Brochure%20IR%202018.pdf) (existantes depuis 2011) car je les trouvent plus complètes et que je ne me penche jamais sur l'impôt avant 2011. En règle général, le code l'impôt sur OpenFisca n'a jamais été vraiment vérifié pour les années anciennes. Je dirais qu'il faut se méfier du calcul de l'IR pour les années avant 2010.

**Comment by sandcha** [comment]
@fpagnoux Que penses-tu de la proposition de Morendil ci-dessus ? 🙂
